### PR TITLE
[MM-36995] Start using db native JSON datatypes

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR converts several DB columns to use native JSON datatypes (`JSON` on MySQL and `jsonb` on PostgreSQL).
This should not only improve application logic and performance when working with the stored JSON data but also remove the size limitations some of these columns had.

I am planning to run a full load-test comparison to verify overall performance hasn't been altered.
Moreover I am going to check the impact of the migration process which I expect to be rather significant, especially on MySQL.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-36995

#### Release Note

```release-note
Converted the following db columns to use native JSON datatypes:
- ChannelMembers.NotifyProps
- Jobs.Data
- LinkMetadata.Data
- Posts.Props
- Sessions.Props
- Threads.Participants
- Users.NotifyProps
- Users.Props
- Users.Timezone
```

### CODE REVIEW

File: store/sqlstore/channel_store.go [hunk 1]
```diff
@@ -1896,45 +1896,27 @@ func (s SqlChannelStore) GetMemberCountsByGroup(ctx context.Context, channelID s
...
-					THEN
-						SUBSTRING(
-						Timezone
-						FROM ` + manualTimezone + `
-						FOR ` + manualTimezoneEnd + `
-					)
+					CASE WHEN Timezone->>"$.useAutomaticTimezone" = 'true' AND LENGTH(Timezone->>"$.automaticTimezone") > 0
+					THEN Timezone->>"$.automaticTimezone"
+					WHEN Timezone->>"$.useAutomaticTimezone" = 'false' AND LENGTH(Timezone->>"$.manualTimezone") > 0
+					THEN Timezone->>"$.manualTimezone"
```
Comment (streamer45) [inline store/sqlstore/channel_store.go:1906]: Had a quick attempt at this. There may be a shorter/cleaner solution just tried to keep the behaviour intact. Let me know if you think of anything better.

File: store/storetest/channel_store.go [hunk 2]
```diff
@@ -4610,8 +4610,11 @@ func testGetMemberCountsByGroup(t *testing.T, ss store.Store) {
 		} else if i == 3 || i == 4 {
 			timeZone["manualTimezone"] = "PST"
 			timeZone["useAutomaticTimezone"] = "false"
-		} else if i == 5 || i == 6 {
-			timeZone["autoTimezone"] = "PST"
+		} else if i == 5 {
+			timeZone["autoTimezone"] = "CET"
+			timeZone["useAutomaticTimezone"] = "true"
+		} else if i == 6 {
+			timeZone["automaticTimezone"] = "CET"
```
Comment (streamer45) [inline store/storetest/channel_store.go:4617]: I *think* the previous implementation had a bug as there were only 2 distinct (and active) timezones but the query returned 3.

File: build/docker-compose.common.yml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 version: '2.4'
 services:
   mysql:
-    image: "mysql:5.6"
+    image: "mysql:5.7"
```
Comment (streamer45) [inline build/docker-compose.common.yml:4]: This is to make tests works on CI.

File: store/sqlstore/channel_store.go (hunk 1 above)
Comment (agnivade) [reply store/sqlstore/channel_store.go:1906]: Oh dear me .. amazed to see hacks like this existed. I somehow missed this :)

Comment (streamer45) [reply store/sqlstore/channel_store.go:1906]: Yeah, it wasn't pretty, that's for sure.

File: store/storetest/channel_store.go [hunk 4]
```diff
@@ -4610,8 +4610,11 @@ func testGetMemberCountsByGroup(t *testing.T, ss store.Store) {
 		} else if i == 3 || i == 4 {
 			timeZone["manualTimezone"] = "PST"
 			timeZone["useAutomaticTimezone"] = "false"
-		} else if i == 5 || i == 6 {
-			timeZone["autoTimezone"] = "PST"
+		} else if i == 5 {
+			timeZone["autoTimezone"] = "CET"
```
Comment (agnivade) [inline store/storetest/channel_store.go:4614]: To clarify: we are deliberately setting `autoTimezone` instead of `automaticTimezone` to test for failure?

Comment (streamer45) [reply store/storetest/channel_store.go:4614]: Yes. I am thinking that was the intent in the previous test so I kept it.

**Review by agnivade — APPROVED** [review]
Looks good. Looking forward to the load-test results.

**Review by mkraft — APPROVED** [review]
Looks great. One question, did you confirm that the existing tests cover each of the fields that are now using JSON/JSONB?

**Comment by streamer45** [comment]
> Looks great. One question, did you confirm that the existing tests cover each of the fields that are now using JSON/JSONB?

From the application side everything should be compatible thanks to the [type converter](https://github.com/mattermost/mattermost-server/blob/38aac770b5950516863f0b719b493b15791845ef/store/sqlstore/store.go#L1569). The only place where I can see things failing would be in the store layer if we somehow query the inner data (like in the timezone case). I couldn't spot any additional occurrence apart from that one. Of course we'll want to run extensive QA on this to make sure everything is working properly.

**Comment by streamer45** [comment]
@furqanmlk Again this is a low level change on the db schema that shouldn't affect overall functionality. I think it would be good to run full e2e tests on this.

**Comment by streamer45** [comment]
#### Performance Review

Ran bounded and unbounded load-tests on a 13M posts dataset (both dbs). Overall performance looks unvaried.
Unbounded load-test on MySQL showed a slight decrease in supported users count (~9%) but it's most likely not related to the changes (see [MM-37100](https://mattermost.atlassian.net/browse/MM-37100)) as bounded load-test looked okay.

```
==================================================
Comparison result:
Grafana Dashboard: http://34.204.52.220:3000/d/6jFqfcZnz/comparison-0-mysql-bounded
master:
  Type: bounded
  DB Engine: mysql
  Duration: 1h
  Users: 2000
  Errors: 38
mm36995:
  Type: bounded
  DB Engine: mysql
  Duration: 1h
  Users: 2000
  Errors: 10
==================================================

==================================================
Comparison result:
Grafana Dashboard: http://54.167.72.59:3000/d/i3KqBcW7k/comparison-0-postgresql-bounded
master:
  Type: bounded
  DB Engine: postgresql
  Duration: 1h
  Users: 3000
  Errors: 132
mm36995:
  Type: bounded
  DB Engine: postgresql
  Duration: 1h
  Users: 3000
  Errors: 134
==================================================

==================================================
Comparison result:
Grafana Dashboard: http://54.196.221.173:3000/d/vgcqfcWnz/comparison-0-postgresql-unbounded
master:
  Type: unbounded
  DB Engine: postgresql
  Supported Users: 7932
  Errors: 1329
mm36995:
  Type: unbounded
  DB Engine: postgresql
  Supported Users: 7926
  Errors: 1829
==================================================

==================================================
Comparison result:
Grafana Dashboard: http://3.80.67.78:3000/d/_mA3B5Z7z/comparison-0-mysql-unbounded
master:
  Type: unbounded
  DB Engine: mysql
  Supported Users: 7872
  Errors: 26310
mm36995:
  Type: unbounded
  DB Engine: mysql
  Supported Users: 7192
  Errors: 23247
==================================================
```

While I was there I've also tried to run a previous version (tested on 5.37) on MySQL and checked that the schema changes are not causing issues.

**Comment by mkraft** [comment]
> #### Performance Review
> Ran bounded and unbounded load-tests on a 13M posts dataset (both dbs). Overall performance looks unvaried.
> Unbounded load-test on MySQL showed a slight decrease in supported users count (~9%) but it's most likely not related to the changes (see [MM-37100](https://mattermost.atlassian.net/browse/MM-37100)) as bounded load-test looked okay.
> 
> ```
> ==================================================
> Comparison result:
> Grafana Dashboard: http://34.204.52.220:3000/d/6jFqfcZnz/comparison-0-mysql-bounded
> master:
>   Type: bounded
>   DB Engine: mysql
>   Duration: 1h
>   Users: 2000
>   Errors: 38
> mm36995:
>   Type: bounded
>   DB Engine: mysql
>   Duration: 1h
>   Users: 2000
>   Errors: 10
> ==================================================
> 
> ==================================================
> Comparison result:
> Grafana Dashboard: http://54.167.72.59:3000/d/i3KqBcW7k/comparison-0-postgresql-bounded
> master:
>   Type: bounded
>   DB Engine: postgresql
>   Duration: 1h
>   Users: 3000
>   Errors: 132
> mm36995:
>   Type: bounded
>   DB Engine: postgresql
>   Duration: 1h
>   Users: 3000
>   Errors: 134
> ==================================================
> 
> ==================================================
> Comparison result:
> Grafana Dashboard: http://54.196.221.173:3000/d/vgcqfcWnz/comparison-0-postgresql-unbounded
> master:
>   Type: unbounded
>   DB Engine: postgresql
>   Supported Users: 7932
>   Errors: 1329
> mm36995:
>   Type: unbounded
>   DB Engine: postgresql
>   Supported Users: 7926
>   Errors: 1829
> ==================================================
> 
> ==================================================
> Comparison result:
> Grafana Dashboard: http://3.80.67.78:3000/d/_mA3B5Z7z/comparison-0-mysql-unbounded
> master:
>   Type: unbounded
>   DB Engine: mysql
>   Supported Users: 7872
>   Errors: 26310
> mm36995:
>   Type: unbounded
>   DB Engine: mysql
>   Supported Users: 7192
>   Errors: 23247
> ==================================================
> ```
> 
> While I was there I've also tried to run a previous version (tested on 5.37) on MySQL and checked that the schema changes are not causing issues.

I wonder if we require an indexing strategy on the value to see any performance gains, for example as outlined here: https://dev.mysql.com/doc/refman/5.7/en/create-table-secondary-indexes.html#json-column-indirect-index

**Comment by agnivade** [comment]
Looks great! Can we also get an estimate of the time it takes for the schema changes for customers to be aware of?

> I wonder if we require an indexing strategy on the value to see any performance gains, for example as outlined here: https://dev.mysql.com/doc/refman/5.7/en/create-table-secondary-indexes.html#json-column-indirect-index

We are just going with a basic schema change to JSON right now. The next phase of the plan is to start looking into optimizing the queries, and depending on the workload, add more schema changes as necessary.

**Comment by streamer45** [comment]
> I wonder if we require an indexing strategy on the value to see any performance gains, for example as outlined here: https://dev.mysql.com/doc/refman/5.7/en/create-table-secondary-indexes.html#json-column-indirect-index

I am thinking we could consider more advanced techniques like that one in the future. Right now we are not using JSON functionalities at all (with the small exception of the user timezone which is not a very stressed field) so I wasn't expecting any performance improvement out of this. I'd say adding indexes at this point would be a bit premature. But overall I agree, it's something we should keep in mind when we start adding more logic that relies on this new data type.

**Comment by streamer45** [comment]
> Looks great! Can we also get an estimate of the time it takes for the schema changes for customers to be aware of?

I've collected some initial numbers but I am thinking of deferring this and do a more in-depth analysis on the whole 6.0 migration at some point in the next few weeks.

**Comment by agnivade** [comment]
Sounds perfect.

Last question - all this will still work with the old codebase right? (For community and community-release).

**Comment by streamer45** [comment]
> Last question - all this will still work with the old codebase right? (For community and community-release).

I've tested 5.37 on MySQL for a bit and everything looked fine. Both insertion and retrieval of JSON fields worked flawlessly.

**Comment by agnivade** [comment]
Gentle ping on this @furqanmlk !

**Comment by furqanmlk** [comment]
> Gentle ping on this @furqanmlk !

Hi @agnivade 
E2E tests are running. 
I will update as soon as they are done

**Review by furqanmlk — APPROVED** [review]
E2E test passed

**Pull request merged**